README.md
# PhonenumberJp

Phonenumber hypenates Japanese phone number.
Based on http://www.soumu.go.jp/main_sosiki/joho_tsusin/top/tel_number/number_shitei.html

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'phonenumber_jp'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install phonenumber_jp

## Usage

    PhonenumberJp.hyphenate("0300000000") # "03-0000-0000"
    PhonenumberJp.hyphenate("0520000000") # "052-000-0000"
    PhonenumberJp.hyphenate("+81300000000") # "+81-3-0000-0000"

## Development

After checking out the repo, run `bin/setup` to install dependencies. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/labocho/phonenumber.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
# v0.1.3 (2021-10-29)

* update special number formats (e.g. 14 digits data transfer number).

# v0.1.2 (2015-07-24)

* copy entire codes from [labocho/phonenumber](https://github.com/labocho/phonenumber).
* add `PhonenumberJp.valid?`

# v0.1.1 (2015-07-24)

* fix gemspec.

# v0.1.0 (2015-07-24)

* initial release.


spec/spec_helper.rb
$LOAD_PATH.unshift(File.expand_path("#{__dir__}/../lib"))
require "phonenumber_jp"

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # This option will default to `:apply_to_host_groups` in RSpec 4 (and will
  # have no way to turn it off -- the option exists only for backwards
  # compatibility in RSpec 3). It causes shared context metadata to be
  # inherited by the metadata hash of host groups and examples, rather than
  # triggering implicit auto-inclusion in groups with matching metadata.
  config.shared_context_metadata_behavior = :apply_to_host_groups

  # The settings below are suggested to provide a good initial experience
  # with RSpec, but feel free to customize to your heart's content.
  #   # This allows you to limit a spec run to individual examples or groups
  #   # you care about by tagging them with `:focus` metadata. When nothing
  #   # is tagged with `:focus`, all examples get run. RSpec also provides
  #   # aliases for `it`, `describe`, and `context` that include `:focus`
  #   # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
  #   config.filter_run_when_matching :focus
  #
  #   # Allows RSpec to persist some state between runs in order to support
  #   # the `--only-failures` and `--next-failure` CLI options. We recommend
  #   # you configure your source control system to ignore this file.
  #   config.example_status_persistence_file_path = "spec/examples.txt"
  #
  #   # Limits the available syntax to the non-monkey patched syntax that is
  #   # recommended. For more details, see:
  #   #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
  #   #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
  #   config.disable_monkey_patching!
  #
  #   # This setting enables warnings. It's recommended, but in some cases may
  #   # be too noisy due to issues in dependencies.
  #   config.warnings = true
  #
  #   # Many RSpec users commonly either run the entire suite or an individual
  #   # file, and it's useful to allow more verbose output when running an
  #   # individual spec file.
  #   if config.files_to_run.one?
  #     # Use the documentation formatter for detailed output,
  #     # unless a formatter has already been configured
  #     # (e.g. via a command-line flag).
  #     config.default_formatter = "doc"
  #   end
  #
  #   # Print the 10 slowest examples and example groups at the
  #   # end of the spec run, to help surface which specs are running
  #   # particularly slow.
  #   config.profile_examples = 10
  #
  #   # Run specs in random order to surface order dependencies. If you find an
  #   # order dependency and want to debug it, you can fix the order by providing
  #   # the seed, which is printed after each run.
  #   #     --seed 1234
  #   config.order = :random
  #
  #   # Seed global randomization in this process using the `--seed` CLI option.
  #   # Setting this allows you to use `--seed` to deterministically reproduce
  #   # test failures related to randomization by passing the same `--seed` value
  #   # as the one that triggered the failure.
  #   Kernel.srand config.seed
end


spec/phonenumber_jp_spec.rb
describe PhonenumberJp do
  describe "VERSION" do
    subject { PhonenumberJp::VERSION }
    it { should be_a String }
  end

  describe ".hyphenate" do
    subject { PhonenumberJp.hyphenate(val) }

    context "nil" do
      let(:val) { nil }
      it { should eq nil }
    end

    context "IP phone" do
      let(:val) { "05012345678" }
      it { should eq "050-1234-5678" }
    end

    context "Free dial (0800)" do
      let(:val) { "08001234567" }
      it { should eq "0800-123-4567" }
    end

    context "Cellular phone" do
      let(:val) { "08012345678" }
      it { should eq "080-1234-5678" }
    end

    context "14 digit data transfer number" do
      let(:val) { "02001234567890" }
      it { should eq "0200-12345-67890" }
    end

    context "Pocket bell" do
      let(:val) { "02042345678" }
      it { should eq "020-423-45678" }
    end

    context "Free dial (0120)" do
      let(:val) { "0120123456" }
      it { should eq "0120-123-456" }
    end

    context "Navi dial" do
      let(:val) { "0570123456" }
      it { should eq "0570-123-456" }
    end

    context "Dial Q2" do
      let(:val) { "0990123456" }
      it { should eq "0990-123-456" }
    end

    context "Area code 0594" do
      let(:val) { "0594123456" }
      it { should eq "0594-12-3456" }
    end

    context "Area code 059" do
      let(:val) { "0592123456" }
      it { should eq "059-212-3456" }
    end

    context "+81" do
      let(:val) { "+81594123456" }
      it { should eq "+81-594-12-3456" }
    end

    context "not match" do
      let(:val) { "9999999999" }
      it { should eq "9999999999" }
    end
  end

  describe "valid?" do
    subject { PhonenumberJp.valid?(val) }

    context "Area code 059" do
      let(:val) { "0592123456" }
      it { should eq true }
    end

    context "+81" do
      let(:val) { "+81594123456" }
      it { should eq true }
    end

    context "not match" do
      let(:val) { "9999999999" }
      it { should eq false }
    end

    context "invalid local" do
      let(:val) { "+81000000000" }
      it { should eq false }
    end
  end
end


scripts/update.rb
# encoding: UTF-8
require "nokogiri"
require "open-uri"
require "uri"
require "tempfile"
require "spreadsheet"
require "csv"
require "date"

include FileUtils

INDEX_URL = "http://www.soumu.go.jp/main_sosiki/joho_tsusin/top/tel_number/number_shitei.html"
EXCEL_FILE_URLS = %w(
  http://www.soumu.go.jp/main_content/000124070.xls
  http://www.soumu.go.jp/main_content/000124071.xls
  http://www.soumu.go.jp/main_content/000124072.xls
  http://www.soumu.go.jp/main_content/000124073.xls
  http://www.soumu.go.jp/main_content/000124074.xls
  http://www.soumu.go.jp/main_content/000124075.xls
  http://www.soumu.go.jp/main_content/000124076.xls
  http://www.soumu.go.jp/main_content/000124077.xls
  http://www.soumu.go.jp/main_content/000124078.xls
)
HEADERS = %w(
  番号区画コード
  MA
  番号
  市外局番
  市内局番
  指定事業者
  使用状況
  備考
)
DATA_DIR = "#{__dir__}/../tmp/data"
CSV_FILENAME = "combined.csv"
REVISION_FILENAME = "REVISION"

def parse_revision(string)
  string = string.strip
  unless string =~ /\A\(平成(\d+)年(\d+)月(\d+)日現在\)\z/
    STDERR.puts "Invalid upated_at: #{string}"
    exit 1
  end
  heisei_y, m, d = $~.captures.map(&:to_i)
  y = heisei_y + 1988
  Date.new(y, m, d)
end

# Checking whether linked from index
index = Nokogiri.parse open(INDEX_URL){|f| f.read }
urls = index.css("a[href]").map{|e| URI.join(INDEX_URL, e["href"]).to_s }
unlinked = EXCEL_FILE_URLS - urls
unless unlinked.empty?
  unlinked.each{|url| STDERR.puts "#{url} is not linked from index" }
  exit 1
end

csv = CSV.open(File.join(DATA_DIR, CSV_FILENAME), "w")
csv << HEADERS
revision = nil

# Download xls
EXCEL_FILE_URLS.each{|url|
  # Download xls
  xls = Tempfile.new("phonenumber")
  STDERR.puts "Download #{url}"
  open(url){|src| xls.write(src.read) }
  xls.close
  # Read xls
  book = Spreadsheet.open(xls.path)
  sheet = book.worksheet(0)
  STDERR.puts "Convert #{url}"
  sheet.rows.each_with_index do |row, index|
    case index
    when 0
      current_revision = parse_revision row.to_a.reject{|cell| cell.nil? || cell.strip == "" }.first
      revision ||= current_revision
      unless revision == current_revision
        STDERR.puts "revision unmatched: #{revision} #{current_revision}"
        exit 1
      end
    when 1
      unless row.to_a == HEADERS
        STDERR.puts "Unexpected header: #{row.to_a.inspect}"
        exit 1
      end
    else
      csv << row.to_a
    end
  end
}

File.write(File.join(DATA_DIR, REVISION_FILENAME), revision.iso8601 + "\n")


lib/phonenumber_jp.rb
require "phonenumber_jp/version"

module PhonenumberJp
  AREA_CODES = File.read("#{__dir__}/phonenumber_jp/area_codes_for_matching.csv").strip.split("\n").map(&:freeze).freeze

  class << self
    # @param phonenumber [String, nil]
    # @return [String, nil]
    def hyphenate(phonenumber)
      parts = split(phonenumber)
      return nil if parts.empty?

      parts.join("-")
    end

    # @param phonenumber [String, nil]
    # @return [true, false]
    def valid?(phonenumber)
      split(phonenumber).length > 1
    end

    private
    # @param phonenumber [String, nil]
    # @return [Array[String]]
    def split(phonenumber)
      return [] if phonenumber.nil?

      phonenumber = phonenumber.gsub(/[^0-9+]/, "")
      return [] if phonenumber.length == 0

      case phonenumber
      when /^\+81/
        local = split("0" + phonenumber[3..-1].to_s) # steep does not know that `phonenumber[3..-1]` will be return String always. Call `to_s` to tell it is String.
        return [phonenumber] if local.length < 2

        local[0] = local[0][1..-1].to_s # steep does not know that `local[0][1..-1]` will be return String always. Call `to_s` to tell it is String.
        ["+81"] + local
      when /^(0120)(\d\d\d)(\d\d\d)$/, # 着信課金機能 (フリーダイヤル)
          /^(0800)(\d\d\d)(\d\d\d\d)$/, # 着信課金機能(フリーダイヤル)
          /^(0180)(\d\d\d)(\d\d\d)$/, # 大量呼受付機能
          /^(0570)(\d\d\d)(\d\d\d)$/, # ナビダイヤル
          /^(0990)(\d\d\d)(\d\d\d)$/, # ダイヤル Q2
          /^(020)(\d\d\d)(\d\d\d\d\d)$/, # データ伝送携帯電話番号, 無線呼出番号
          /^(0200)(\d\d\d\d\d)(\d\d\d\d\d)$/, # データ伝送携帯電話番号
          /^(0[789]0)(\d\d\d\d)(\d\d\d\d)$/, # 携帯電話・PHS
          /^(050)(\d\d\d\d)(\d\d\d\d)$/ # IP 電話
        $~.captures
      when /^0(\d{9})$/ # 固定電話
        return [phonenumber] unless (area_code = find_area_code(phonenumber))

        l = area_code.length
        [
          phonenumber[0..(l - 1)],
          phonenumber[l..5],
          phonenumber[6..-1],
        ]
      else
        [phonenumber]
      end
    end

    # @param phonenumber [String]
    # @return [String, nil]
    def find_area_code(phonenumber)
      AREA_CODES.find {|area_code|
        phonenumber[0..(area_code.length - 1)] == area_code
      }
    end
  end
end
